storage: HeapSort CommandQueue overlaps to permit early sort termination

nvb opened the pull request:
We previously introduced an optimization where we could cut dependency
searching in `CommandQueue.GetWait` short once the established
dependencies fully enclosed the new command. This opened an opportunity
to avoid sorting the entire overlapping command slice and instead
sorting on demand using a Heap. As @bdarnell mentioned:

> "You don't need to know the top K in advance or do any chunking; just
> make the whole overlaps array into a heap (O(n)), then peel off one
> element at a time (O(log(n))). If you do need the entire array then
> it's the same O(n*log(n)) of the Sort() call, although with a somewhat
> higher constant factor."

This PR makes this proposed change and hopes to serve as a discussion
about its performance impacts.
#### `GetWait` Dependency Calculation Stats:

The following metrics were gathered by instrumenting `GetWait` with
`go-metric` and hitting a single node with a photos example app.
This was meant to be a representative example of semi-contentious load.
The metrics gathered were:
- `pre overlap`: the number of overlapping commands
- `post overlap`: the number of commands left in the heap after
  calculating dependencies. These are the commands that we avoided
  sorting
- `percent bailed`: the percent of commands that we avoided sorting by
  using an online heapsort
##### 50 users:

```
percent bailed.50-percentile: 67,
percent bailed.75-percentile: 76,
percent bailed.95-percentile: 89,
percent bailed.99-percentile: 98,
percent bailed.999-percentile: 99,
percent bailed.count: 70215,
percent bailed.max: 99,
percent bailed.mean: 49.1632080078125,
percent bailed.min: 0,
percent bailed.std-dev: 34.15071502588194,
post overlap length.50-percentile: 2,
post overlap length.75-percentile: 4,
post overlap length.95-percentile: 10,
post overlap length.99-percentile: 60,
post overlap length.999-percentile: 83,
post overlap length.count: 70215,
post overlap length.max: 96,
post overlap length.mean: 3.9793701171875,
post overlap length.min: 0,
post overlap length.std-dev: 9.756680363352212,
pre overlap length.50-percentile: 3,
pre overlap length.75-percentile: 5,
pre overlap length.95-percentile: 33,
pre overlap length.99-percentile: 69,
pre overlap length.999-percentile: 91,
pre overlap length.count: 70215,
pre overlap length.max: 97,
pre overlap length.mean: 6.13134765625,
pre overlap length.min: 1,
pre overlap length.std-dev: 12.64766621491185
```
##### 1000 users:

```
percent bailed.50-percentile: 80,
percent bailed.75-percentile: 91,
percent bailed.95-percentile: 99,
percent bailed.99-percentile: 100,
percent bailed.999-percentile: 100,
percent bailed.count: 19316,
percent bailed.max: 100,
percent bailed.mean: 68.78369140625,
percent bailed.min: 0,
percent bailed.std-dev: 33.37719207522773,
post overlap length.50-percentile: 4,
post overlap length.75-percentile: 10,
post overlap length.95-percentile: 174,
post overlap length.99-percentile: 1063.0699999999997,
post overlap length.999-percentile: 1482.4209999999994,
post overlap length.count: 19316,
post overlap length.max: 1496,
post overlap length.mean: 43.474609375,
post overlap length.min: 0,
post overlap length.std-dev: 161.09362296868096,
pre overlap length.50-percentile: 6,
pre overlap length.75-percentile: 11,
pre overlap length.95-percentile: 175,
pre overlap length.99-percentile: 1051,
pre overlap length.999-percentile: 1473.6139999999996,
pre overlap length.count: 19316,
pre overlap length.max: 1497,
pre overlap length.mean: 44.2047119140625,
pre overlap length.min: 1,
pre overlap length.std-dev: 157.83892248564868
```
#### Heapsort vs. Quicksort Performance:

Benchmarking the old and new sorting algorithms independently. Notice
that a major issue with heapsort is that `heap.Pop` allocates, so the
number of allocations is linear with respect to the number of
dependencies we search. Also notice that it is roughly twice as slow to
sort the same size slice, which coincidentally lines with the fact that
from the above stats, it looks like we avoid sorting about half of the
elements.
##### 97 elements (max overlaps from above):

```
BenchmarkQsort-4     1000000          1087 ns/op          32 B/op          1 allocs/op
BenchmarkHeap-4      1000000          1619 ns/op          80 B/op          7 allocs/op
```
##### 6 elements (average overlaps from above):

```
BenchmarkQsort-4      200000          9037 ns/op          32 B/op          1 allocs/op
BenchmarkHeap-4       100000         16547 ns/op         816 B/op         98 allocs/op
```
##### 1497 elements (max overlaps from 1000 users above):

```
BenchmarkQsort-4       10000        178428 ns/op          32 B/op          1 allocs/op
BenchmarkHeap-4         5000        342415 ns/op       12016 B/op       1498 allocs/op
```
##### 6 elements, only pop 2 for heapsort:

```
BenchmarkQsort-4     1000000          1045 ns/op          32 B/op          1 allocs/op
BenchmarkHeap-4      1000000          1284 ns/op          48 B/op          3 allocs/op
```
##### 44 elements, only pop 15 for heapsort:

```
BenchmarkQsort-4      300000          4255 ns/op          32 B/op          1 allocs/op
BenchmarkHeap-4       300000          4619 ns/op         160 B/op         16 allocs/op
```
#### SQL Benchmarks:

```
name                          old time/op    new time/op    delta
Bank_Cockroach-2                 421µs ± 2%     424µs ± 1%    ~           (p=0.089 n=10+10)
Select1_Cockroach-2             60.0µs ± 1%    60.4µs ± 1%    ~           (p=0.052 n=10+10)
Select2_Cockroach-2             1.03ms ± 2%    1.02ms ± 1%    ~           (p=0.436 n=10+10)
Insert1_Cockroach-2              500µs ± 1%     500µs ± 1%    ~           (p=0.796 n=10+10)
Insert10_Cockroach-2             887µs ± 1%     886µs ± 1%    ~            (p=0.447 n=10+9)
Insert100_Cockroach-2           4.51ms ± 1%    4.48ms ± 1%  -0.60%        (p=0.023 n=10+10)
Insert1000_Cockroach-2          43.7ms ± 1%    43.4ms ± 1%    ~           (p=0.052 n=10+10)
Update1_Cockroach-2              757µs ± 2%     752µs ± 1%    ~            (p=0.122 n=8+10)
Update10_Cockroach-2            1.40ms ± 1%    1.40ms ± 2%    ~            (p=0.905 n=9+10)
Update100_Cockroach-2           7.54ms ± 3%    7.51ms ± 2%    ~           (p=0.579 n=10+10)
Update1000_Cockroach-2          66.8ms ± 1%    66.9ms ± 3%    ~            (p=0.661 n=9+10)
Delete1_Cockroach-2              844µs ± 2%     842µs ± 2%    ~           (p=0.739 n=10+10)
Delete10_Cockroach-2            2.62ms ± 5%    2.59ms ± 4%    ~           (p=0.190 n=10+10)
Delete100_Cockroach-2           21.1ms ± 1%    21.2ms ± 2%    ~            (p=0.497 n=10+9)
Delete1000_Cockroach-2           222ms ± 2%     223ms ± 2%    ~           (p=0.796 n=10+10)
Scan1_Cockroach-2                268µs ± 1%     268µs ± 1%    ~            (p=0.762 n=8+10)
Scan10_Cockroach-2               315µs ± 1%     318µs ± 2%    ~           (p=0.063 n=10+10)
Scan100_Cockroach-2              666µs ± 1%     668µs ± 1%    ~            (p=0.549 n=9+10)
Scan1000_Cockroach-2            3.97ms ± 1%    3.97ms ± 1%    ~           (p=0.853 n=10+10)
Scan10000_Cockroach-2           40.6ms ± 1%    40.9ms ± 1%  +0.87%        (p=0.015 n=10+10)
Scan1000Limit1_Cockroach-2       284µs ± 1%     285µs ± 1%    ~             (p=1.000 n=9+9)
Scan1000Limit10_Cockroach-2      331µs ± 2%     330µs ± 1%    ~           (p=0.684 n=10+10)
Scan1000Limit100_Cockroach-2     677µs ± 1%     686µs ± 3%    ~           (p=0.052 n=10+10)
PgbenchQuery_Cockroach-2        3.95ms ± 5%    3.96ms ± 3%    ~           (p=0.853 n=10+10)

name                          old alloc/op   new alloc/op   delta
Bank_Cockroach-2                68.6kB ± 0%    68.6kB ± 0%    ~            (p=0.858 n=10+9)
Select1_Cockroach-2             3.90kB ± 0%    3.91kB ± 0%  +0.11%        (p=0.016 n=10+10)
Select2_Cockroach-2             80.8kB ± 0%    80.8kB ± 0%    ~            (p=0.371 n=8+10)
Insert1_Cockroach-2             23.1kB ± 0%    23.1kB ± 0%    ~           (p=0.868 n=10+10)
Insert10_Cockroach-2            67.6kB ± 0%    67.5kB ± 0%    ~            (p=0.128 n=9+10)
Insert100_Cockroach-2            512kB ± 0%     512kB ± 0%    ~           (p=1.000 n=10+10)
Insert1000_Cockroach-2          4.61MB ± 1%    4.59MB ± 1%    ~           (p=0.436 n=10+10)
Update1_Cockroach-2             37.0kB ± 0%    37.0kB ± 0%    ~             (p=0.713 n=9+9)
Update10_Cockroach-2             101kB ± 0%     101kB ± 0%    ~            (p=0.633 n=8+10)
Update100_Cockroach-2            712kB ± 0%     712kB ± 0%    ~             (p=0.743 n=9+8)
Update1000_Cockroach-2          6.21MB ± 1%    6.18MB ± 1%    ~            (p=0.156 n=10+9)
Delete1_Cockroach-2             36.1kB ± 0%    36.2kB ± 0%    ~            (p=0.180 n=8+10)
Delete10_Cockroach-2             110kB ± 0%     110kB ± 0%    ~             (p=0.352 n=9+9)
Delete100_Cockroach-2            824kB ± 0%     825kB ± 0%    ~            (p=0.356 n=9+10)
Delete1000_Cockroach-2          7.44MB ± 0%    7.47MB ± 1%    ~            (p=0.237 n=8+10)
Scan1_Cockroach-2               13.3kB ± 0%    13.3kB ± 0%    ~             (p=0.199 n=9+9)
Scan10_Cockroach-2              17.2kB ± 0%    17.2kB ± 0%    ~           (p=0.810 n=10+10)
Scan100_Cockroach-2             49.4kB ± 0%    49.5kB ± 0%    ~            (p=0.387 n=9+10)
Scan1000_Cockroach-2             331kB ± 0%     331kB ± 0%  +0.01%          (p=0.012 n=9+8)
Scan10000_Cockroach-2           5.72MB ± 0%    5.72MB ± 0%    ~            (p=0.986 n=9+10)
Scan1000Limit1_Cockroach-2      14.1kB ± 0%    14.1kB ± 0%    ~           (p=0.593 n=10+10)
Scan1000Limit10_Cockroach-2     17.9kB ± 0%    18.0kB ± 0%  +0.12%          (p=0.021 n=9+8)
Scan1000Limit100_Cockroach-2    50.1kB ± 0%    50.1kB ± 0%    ~             (p=0.531 n=9+9)
PgbenchQuery_Cockroach-2         227kB ±10%     227kB ±10%    ~           (p=0.853 n=10+10)

name                          old allocs/op  new allocs/op  delta
Bank_Cockroach-2                 1.67k ± 0%     1.67k ± 0%    ~            (p=0.294 n=8+10)
Select1_Cockroach-2               97.0 ± 0%      97.0 ± 0%    ~     (all samples are equal)
Select2_Cockroach-2              2.70k ± 0%     2.70k ± 0%    ~            (p=0.612 n=9+10)
Insert1_Cockroach-2                357 ± 0%       357 ± 0%    ~           (p=1.000 n=10+10)
Insert10_Cockroach-2               916 ± 0%       916 ± 0%    ~           (p=0.559 n=10+10)
Insert100_Cockroach-2            6.28k ± 0%     6.28k ± 0%    ~             (p=0.327 n=8+9)
Insert1000_Cockroach-2           60.1k ± 0%     60.0k ± 0%    ~           (p=0.315 n=10+10)
Update1_Cockroach-2                720 ± 0%       720 ± 0%    ~             (p=0.388 n=9+8)
Update10_Cockroach-2             1.43k ± 0%     1.43k ± 0%    ~            (p=0.758 n=8+10)
Update100_Cockroach-2            8.15k ± 0%     8.15k ± 0%    ~             (p=0.516 n=7+8)
Update1000_Cockroach-2           72.4k ± 1%     72.3k ± 0%    ~            (p=0.588 n=10+9)
Delete1_Cockroach-2                693 ± 0%       693 ± 0%    ~            (p=0.500 n=7+10)
Delete10_Cockroach-2             1.72k ± 0%     1.72k ± 0%  -0.06%          (p=0.009 n=9+8)
Delete100_Cockroach-2            11.6k ± 0%     11.6k ± 0%  +0.03%         (p=0.036 n=8+10)
Delete1000_Cockroach-2            110k ± 0%      110k ± 0%    ~            (p=0.166 n=8+10)
Scan1_Cockroach-2                  276 ± 1%       276 ± 1%    ~           (p=0.825 n=10+10)
Scan10_Cockroach-2                 356 ± 0%       355 ± 0%    ~           (p=0.714 n=10+10)
Scan100_Cockroach-2              1.08k ± 0%     1.08k ± 0%    ~            (p=0.506 n=7+10)
Scan1000_Cockroach-2             8.29k ± 0%     8.29k ± 0%    ~     (all samples are equal)
Scan10000_Cockroach-2            80.4k ± 0%     80.4k ± 0%    ~           (p=0.287 n=10+10)
Scan1000Limit1_Cockroach-2         298 ± 0%       298 ± 0%    ~            (p=0.549 n=10+8)
Scan1000Limit10_Cockroach-2        377 ± 0%       378 ± 0%  +0.21%         (p=0.008 n=8+10)
Scan1000Limit100_Cockroach-2     1.10k ± 0%     1.10k ± 0%    ~           (p=1.000 n=10+10)
PgbenchQuery_Cockroach-2         4.05k ± 7%     4.05k ± 7%    ~           (p=0.934 n=10+10)
```
#### Analysis:

All in all, this looks pretty insignificant either way. 99th percentile
performance could be argued as a reason for this change. The worse
constant factor for all sorting could be argued as a reason against it.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/4878)

<!-- Reviewable:end -->

nvb committed:
storage: HeapSort CommandQueue overlaps to permit early breaking

We previously introduced an optimization where we could cut dependency
searching in `CommandQueue.GetWait` short once the established
dependencies fully enclosed the new command. This opened an opportunity
to avoid sorting the entire overlapping command slice and instead
sorting on demand using a Heap. As @bdarnell mentioned:

> "You don't need to know the top K in advance or do any chunking; just
> make the whole overlaps array into a heap (O(n)), then peel off one
> element at a time (O(log(n))). If you do need the entire array then
> it's the same O(n*log(n)) of the Sort() call, although with a somewhat
> higher constant factor."

This PR makes this proposed change and hopes to serve as a discussion
about its performance impacts.

`GetWait` Dependency Calculation Stats:

The following metrics were gathered by instrumenting `GetWait` with
`go-metric` and hitting a single node with a photos example app.
This was meant to be a representative example of semi-contentious load.
The metrics gathered were:
- `pre overlap`: the number of overlapping commands
- `post overlap`: the number of commands left in the heap after
  calculating dependencies. These are the commands that we avoided
  sorting.
- `percent bailed`: the percent of commands that we avoided sorting by
  using an online heapsort.

50 users:
```
percent bailed.50-percentile: 67,
percent bailed.75-percentile: 76,
percent bailed.95-percentile: 89,
percent bailed.99-percentile: 98,
percent bailed.999-percentile: 99,
percent bailed.count: 70215,
percent bailed.max: 99,
percent bailed.mean: 49.1632080078125,
percent bailed.min: 0,
percent bailed.std-dev: 34.15071502588194,
post overlap length.50-percentile: 2,
post overlap length.75-percentile: 4,
post overlap length.95-percentile: 10,
post overlap length.99-percentile: 60,
post overlap length.999-percentile: 83,
post overlap length.count: 70215,
post overlap length.max: 96,
post overlap length.mean: 3.9793701171875,
post overlap length.min: 0,
post overlap length.std-dev: 9.756680363352212,
pre overlap length.50-percentile: 3,
pre overlap length.75-percentile: 5,
pre overlap length.95-percentile: 33,
pre overlap length.99-percentile: 69,
pre overlap length.999-percentile: 91,
pre overlap length.count: 70215,
pre overlap length.max: 97,
pre overlap length.mean: 6.13134765625,
pre overlap length.min: 1,
pre overlap length.std-dev: 12.64766621491185
```

1000 users:
```
percent bailed.50-percentile: 80,
percent bailed.75-percentile: 91,
percent bailed.95-percentile: 99,
percent bailed.99-percentile: 100,
percent bailed.999-percentile: 100,
percent bailed.count: 19316,
percent bailed.max: 100,
percent bailed.mean: 68.78369140625,
percent bailed.min: 0,
percent bailed.std-dev: 33.37719207522773,
post overlap length.50-percentile: 4,
post overlap length.75-percentile: 10,
post overlap length.95-percentile: 174,
post overlap length.99-percentile: 1063.0699999999997,
post overlap length.999-percentile: 1482.4209999999994,
post overlap length.count: 19316,
post overlap length.max: 1496,
post overlap length.mean: 43.474609375,
post overlap length.min: 0,
post overlap length.std-dev: 161.09362296868096,
pre overlap length.50-percentile: 6,
pre overlap length.75-percentile: 11,
pre overlap length.95-percentile: 175,
pre overlap length.99-percentile: 1051,
pre overlap length.999-percentile: 1473.6139999999996,
pre overlap length.count: 19316,
pre overlap length.max: 1497,
pre overlap length.mean: 44.2047119140625,
pre overlap length.min: 1,
pre overlap length.std-dev: 157.83892248564868
```

Heapsort vs. Quicksort Performance:

Benchmarking the old and new sorting algorithms independently. Notice
that a major issue with heapsort is that `heap.Pop` allocates, so the
number of allocations is linear with respect to the number of
dependencies we search. Also notice that it is roughly twice as slow to
sort the same size slice, which coincidentally lines with the fact that
from the above stats, it looks like we avoid sorting about half of the
elements.

97 elements (max overlaps from 50 users above):
```
BenchmarkQsort-4	 1000000	      1087 ns/op	      32 B/op	       1 allocs/op
BenchmarkHeap-4 	 1000000	      1619 ns/op	      80 B/op	       7 allocs/op
```

6 elements (average overlaps from 50 users above):
```
BenchmarkQsort-4	  200000	      9037 ns/op	      32 B/op	       1 allocs/op
BenchmarkHeap-4 	  100000	     16547 ns/op	     816 B/op	      98 allocs/op
```

1497 elements (max overlaps from 1000 users above):
```
BenchmarkQsort-4	   10000	    178428 ns/op	      32 B/op	       1 allocs/op
BenchmarkHeap-4 	    5000	    342415 ns/op	   12016 B/op	    1498 allocs/op
```

6 elements, only pop 2 for heapsort:
```
BenchmarkQsort-4	 1000000	      1045 ns/op	      32 B/op	       1 allocs/op
BenchmarkHeap-4 	 1000000	      1284 ns/op	      48 B/op	       3 allocs/op
```

44 elements, only pop 15 for heapsort:
```
BenchmarkQsort-4	  300000	      4255 ns/op	      32 B/op	       1 allocs/op
BenchmarkHeap-4 	  300000	      4619 ns/op	     160 B/op	      16 allocs/op
```

SQL Benchmarks:

```
name                          old time/op    new time/op    delta
Bank_Cockroach-2                 421µs ± 2%     424µs ± 1%    ~           (p=0.089 n=10+10)
Select1_Cockroach-2             60.0µs ± 1%    60.4µs ± 1%    ~           (p=0.052 n=10+10)
Select2_Cockroach-2             1.03ms ± 2%    1.02ms ± 1%    ~           (p=0.436 n=10+10)
Insert1_Cockroach-2              500µs ± 1%     500µs ± 1%    ~           (p=0.796 n=10+10)
Insert10_Cockroach-2             887µs ± 1%     886µs ± 1%    ~            (p=0.447 n=10+9)
Insert100_Cockroach-2           4.51ms ± 1%    4.48ms ± 1%  -0.60%        (p=0.023 n=10+10)
Insert1000_Cockroach-2          43.7ms ± 1%    43.4ms ± 1%    ~           (p=0.052 n=10+10)
Update1_Cockroach-2              757µs ± 2%     752µs ± 1%    ~            (p=0.122 n=8+10)
Update10_Cockroach-2            1.40ms ± 1%    1.40ms ± 2%    ~            (p=0.905 n=9+10)
Update100_Cockroach-2           7.54ms ± 3%    7.51ms ± 2%    ~           (p=0.579 n=10+10)
Update1000_Cockroach-2          66.8ms ± 1%    66.9ms ± 3%    ~            (p=0.661 n=9+10)
Delete1_Cockroach-2              844µs ± 2%     842µs ± 2%    ~           (p=0.739 n=10+10)
Delete10_Cockroach-2            2.62ms ± 5%    2.59ms ± 4%    ~           (p=0.190 n=10+10)
Delete100_Cockroach-2           21.1ms ± 1%    21.2ms ± 2%    ~            (p=0.497 n=10+9)
Delete1000_Cockroach-2           222ms ± 2%     223ms ± 2%    ~           (p=0.796 n=10+10)
Scan1_Cockroach-2                268µs ± 1%     268µs ± 1%    ~            (p=0.762 n=8+10)
Scan10_Cockroach-2               315µs ± 1%     318µs ± 2%    ~           (p=0.063 n=10+10)
Scan100_Cockroach-2              666µs ± 1%     668µs ± 1%    ~            (p=0.549 n=9+10)
Scan1000_Cockroach-2            3.97ms ± 1%    3.97ms ± 1%    ~           (p=0.853 n=10+10)
Scan10000_Cockroach-2           40.6ms ± 1%    40.9ms ± 1%  +0.87%        (p=0.015 n=10+10)
Scan1000Limit1_Cockroach-2       284µs ± 1%     285µs ± 1%    ~             (p=1.000 n=9+9)
Scan1000Limit10_Cockroach-2      331µs ± 2%     330µs ± 1%    ~           (p=0.684 n=10+10)
Scan1000Limit100_Cockroach-2     677µs ± 1%     686µs ± 3%    ~           (p=0.052 n=10+10)
PgbenchQuery_Cockroach-2        3.95ms ± 5%    3.96ms ± 3%    ~           (p=0.853 n=10+10)

name                          old alloc/op   new alloc/op   delta
Bank_Cockroach-2                68.6kB ± 0%    68.6kB ± 0%    ~            (p=0.858 n=10+9)
Select1_Cockroach-2             3.90kB ± 0%    3.91kB ± 0%  +0.11%        (p=0.016 n=10+10)
Select2_Cockroach-2             80.8kB ± 0%    80.8kB ± 0%    ~            (p=0.371 n=8+10)
Insert1_Cockroach-2             23.1kB ± 0%    23.1kB ± 0%    ~           (p=0.868 n=10+10)
Insert10_Cockroach-2            67.6kB ± 0%    67.5kB ± 0%    ~            (p=0.128 n=9+10)
Insert100_Cockroach-2            512kB ± 0%     512kB ± 0%    ~           (p=1.000 n=10+10)
Insert1000_Cockroach-2          4.61MB ± 1%    4.59MB ± 1%    ~           (p=0.436 n=10+10)
Update1_Cockroach-2             37.0kB ± 0%    37.0kB ± 0%    ~             (p=0.713 n=9+9)
Update10_Cockroach-2             101kB ± 0%     101kB ± 0%    ~            (p=0.633 n=8+10)
Update100_Cockroach-2            712kB ± 0%     712kB ± 0%    ~             (p=0.743 n=9+8)
Update1000_Cockroach-2          6.21MB ± 1%    6.18MB ± 1%    ~            (p=0.156 n=10+9)
Delete1_Cockroach-2             36.1kB ± 0%    36.2kB ± 0%    ~            (p=0.180 n=8+10)
Delete10_Cockroach-2             110kB ± 0%     110kB ± 0%    ~             (p=0.352 n=9+9)
Delete100_Cockroach-2            824kB ± 0%     825kB ± 0%    ~            (p=0.356 n=9+10)
Delete1000_Cockroach-2          7.44MB ± 0%    7.47MB ± 1%    ~            (p=0.237 n=8+10)
Scan1_Cockroach-2               13.3kB ± 0%    13.3kB ± 0%    ~             (p=0.199 n=9+9)
Scan10_Cockroach-2              17.2kB ± 0%    17.2kB ± 0%    ~           (p=0.810 n=10+10)
Scan100_Cockroach-2             49.4kB ± 0%    49.5kB ± 0%    ~            (p=0.387 n=9+10)
Scan1000_Cockroach-2             331kB ± 0%     331kB ± 0%  +0.01%          (p=0.012 n=9+8)
Scan10000_Cockroach-2           5.72MB ± 0%    5.72MB ± 0%    ~            (p=0.986 n=9+10)
Scan1000Limit1_Cockroach-2      14.1kB ± 0%    14.1kB ± 0%    ~           (p=0.593 n=10+10)
Scan1000Limit10_Cockroach-2     17.9kB ± 0%    18.0kB ± 0%  +0.12%          (p=0.021 n=9+8)
Scan1000Limit100_Cockroach-2    50.1kB ± 0%    50.1kB ± 0%    ~             (p=0.531 n=9+9)
PgbenchQuery_Cockroach-2         227kB ±10%     227kB ±10%    ~           (p=0.853 n=10+10)

name                          old allocs/op  new allocs/op  delta
Bank_Cockroach-2                 1.67k ± 0%     1.67k ± 0%    ~            (p=0.294 n=8+10)
Select1_Cockroach-2               97.0 ± 0%      97.0 ± 0%    ~     (all samples are equal)
Select2_Cockroach-2              2.70k ± 0%     2.70k ± 0%    ~            (p=0.612 n=9+10)
Insert1_Cockroach-2                357 ± 0%       357 ± 0%    ~           (p=1.000 n=10+10)
Insert10_Cockroach-2               916 ± 0%       916 ± 0%    ~           (p=0.559 n=10+10)
Insert100_Cockroach-2            6.28k ± 0%     6.28k ± 0%    ~             (p=0.327 n=8+9)
Insert1000_Cockroach-2           60.1k ± 0%     60.0k ± 0%    ~           (p=0.315 n=10+10)
Update1_Cockroach-2                720 ± 0%       720 ± 0%    ~             (p=0.388 n=9+8)
Update10_Cockroach-2             1.43k ± 0%     1.43k ± 0%    ~            (p=0.758 n=8+10)
Update100_Cockroach-2            8.15k ± 0%     8.15k ± 0%    ~             (p=0.516 n=7+8)
Update1000_Cockroach-2           72.4k ± 1%     72.3k ± 0%    ~            (p=0.588 n=10+9)
Delete1_Cockroach-2                693 ± 0%       693 ± 0%    ~            (p=0.500 n=7+10)
Delete10_Cockroach-2             1.72k ± 0%     1.72k ± 0%  -0.06%          (p=0.009 n=9+8)
Delete100_Cockroach-2            11.6k ± 0%     11.6k ± 0%  +0.03%         (p=0.036 n=8+10)
Delete1000_Cockroach-2            110k ± 0%      110k ± 0%    ~            (p=0.166 n=8+10)
Scan1_Cockroach-2                  276 ± 1%       276 ± 1%    ~           (p=0.825 n=10+10)
Scan10_Cockroach-2                 356 ± 0%       355 ± 0%    ~           (p=0.714 n=10+10)
Scan100_Cockroach-2              1.08k ± 0%     1.08k ± 0%    ~            (p=0.506 n=7+10)
Scan1000_Cockroach-2             8.29k ± 0%     8.29k ± 0%    ~     (all samples are equal)
Scan10000_Cockroach-2            80.4k ± 0%     80.4k ± 0%    ~           (p=0.287 n=10+10)
Scan1000Limit1_Cockroach-2         298 ± 0%       298 ± 0%    ~            (p=0.549 n=10+8)
Scan1000Limit10_Cockroach-2        377 ± 0%       378 ± 0%  +0.21%         (p=0.008 n=8+10)
Scan1000Limit100_Cockroach-2     1.10k ± 0%     1.10k ± 0%    ~           (p=1.000 n=10+10)
PgbenchQuery_Cockroach-2         4.05k ± 7%     4.05k ± 7%    ~           (p=0.934 n=10+10)
```

Analysis:

All in all, this looks pretty insignificant either way. 99th percentile
performance could be argued as a reason for this change. The worse
constant factor for all sorting could be argued as a reason against it.

bdarnell commented:
Thanks for running these experiments. If it doesn't make much difference, I think I'd rather stay with the sorting version, mainly to avoid the allocation implied by the heap interface.

nvb committed:
storage: Avoid all HeapSort allocations in CommandQueue.GetWait

`heap.Pop` returns an `interface{}`, meaning that it allocates on every
pop for return values. However, by returning a pointer, we can avoid all
allocations when popping. We can then pull the popped value out of the pointer.

Effect on Heap Benchmarks:

```
name    old time/op    new time/op    delta
Heap-4     166µs ± 1%     147µs ± 1%  -11.43%  (p=0.000 n=10+10)

name    old alloc/op   new alloc/op   delta
Heap-4    6.43kB ± 0%    0.03kB ± 0%  -99.50%  (p=0.000 n=10+10)

name    old allocs/op  new allocs/op  delta
Heap-4       801 ± 0%         1 ± 0%  -99.88%  (p=0.000 n=10+10)
```

New Perf Impact of PR:

```
name                          old time/op    new time/op    delta
Bank_Cockroach-2                 424µs ± 3%     424µs ± 1%    ~           (p=0.684 n=10+10)
Select1_Cockroach-2             60.1µs ± 1%    60.3µs ± 1%    ~           (p=0.247 n=10+10)
Select2_Cockroach-2             1.03ms ± 1%    1.03ms ± 2%    ~           (p=0.579 n=10+10)
Insert1_Cockroach-2              502µs ± 1%     498µs ± 1%  -0.87%        (p=0.000 n=10+10)
Insert10_Cockroach-2             890µs ± 1%     885µs ± 1%  -0.65%        (p=0.023 n=10+10)
Insert100_Cockroach-2           4.51ms ± 1%    4.51ms ± 1%    ~            (p=0.780 n=9+10)
Insert1000_Cockroach-2          43.7ms ± 1%    43.7ms ± 1%    ~           (p=0.684 n=10+10)
Update1_Cockroach-2              763µs ± 1%     758µs ± 1%  -0.63%         (p=0.006 n=10+9)
Update10_Cockroach-2            1.41ms ± 2%    1.40ms ± 1%    ~           (p=0.075 n=10+10)
Update100_Cockroach-2           7.53ms ± 2%    7.53ms ± 2%    ~           (p=0.912 n=10+10)
Update1000_Cockroach-2          68.1ms ± 2%    68.0ms ± 2%    ~           (p=0.684 n=10+10)
Delete1_Cockroach-2              843µs ± 1%     839µs ± 2%    ~            (p=0.237 n=8+10)
Delete10_Cockroach-2            2.58ms ± 2%    2.58ms ± 3%    ~            (p=0.968 n=9+10)
Delete100_Cockroach-2           21.4ms ± 1%    21.2ms ± 2%  -1.02%         (p=0.043 n=8+10)
Delete1000_Cockroach-2           224ms ± 2%     223ms ± 2%    ~           (p=0.853 n=10+10)
Scan1_Cockroach-2                270µs ± 1%     267µs ± 1%  -0.86%        (p=0.000 n=10+10)
Scan10_Cockroach-2               317µs ± 2%     312µs ± 1%  -1.44%        (p=0.000 n=10+10)
Scan100_Cockroach-2              671µs ± 3%     668µs ± 4%    ~           (p=0.481 n=10+10)
Scan1000_Cockroach-2            3.95ms ± 2%    3.94ms ± 1%    ~           (p=0.315 n=10+10)
Scan10000_Cockroach-2           40.8ms ± 1%    40.8ms ± 3%    ~           (p=0.853 n=10+10)
Scan1000Limit1_Cockroach-2       285µs ± 1%     283µs ± 1%  -0.68%        (p=0.029 n=10+10)
Scan1000Limit10_Cockroach-2      330µs ± 1%     328µs ± 1%    ~            (p=0.278 n=10+9)
Scan1000Limit100_Cockroach-2     692µs ± 2%     687µs ± 4%    ~           (p=0.165 n=10+10)
PgbenchQuery_Cockroach-2        3.90ms ± 5%    3.84ms ± 2%    ~             (p=0.063 n=9+9)

name                          old alloc/op   new alloc/op   delta
Bank_Cockroach-2                68.6kB ± 0%    68.6kB ± 0%    ~           (p=0.110 n=10+10)
Select1_Cockroach-2             3.90kB ± 0%    3.91kB ± 0%    ~           (p=0.182 n=10+10)
Select2_Cockroach-2             80.8kB ± 0%    80.8kB ± 0%  -0.02%          (p=0.049 n=9+8)
Insert1_Cockroach-2             23.1kB ± 0%    23.0kB ± 0%  -0.35%         (p=0.001 n=9+10)
Insert10_Cockroach-2            67.6kB ± 0%    67.2kB ± 0%  -0.61%        (p=0.000 n=10+10)
Insert100_Cockroach-2            511kB ± 0%     508kB ± 0%  -0.62%         (p=0.000 n=9+10)
Insert1000_Cockroach-2          4.61MB ± 1%    4.58MB ± 1%  -0.81%        (p=0.001 n=10+10)
Update1_Cockroach-2             37.0kB ± 0%    36.9kB ± 0%  -0.39%        (p=0.000 n=10+10)
Update10_Cockroach-2             101kB ± 0%     100kB ± 0%  -0.72%         (p=0.000 n=9+10)
Update100_Cockroach-2            712kB ± 0%     710kB ± 2%    ~           (p=0.143 n=10+10)
Update1000_Cockroach-2          6.21MB ± 0%    6.12MB ± 2%  -1.33%          (p=0.012 n=6+9)
Delete1_Cockroach-2             36.1kB ± 0%    36.0kB ± 0%  -0.23%         (p=0.001 n=10+8)
Delete10_Cockroach-2             110kB ± 0%     109kB ± 0%  -0.62%          (p=0.000 n=8+9)
Delete100_Cockroach-2            825kB ± 0%     818kB ± 0%  -0.78%          (p=0.000 n=9+9)
Delete1000_Cockroach-2          7.44MB ± 0%    7.37MB ± 0%  -0.87%          (p=0.000 n=7+8)
Scan1_Cockroach-2               13.3kB ± 0%    13.3kB ± 0%  -0.23%          (p=0.015 n=9+9)
Scan10_Cockroach-2              17.2kB ± 0%    17.2kB ± 0%    ~           (p=0.110 n=10+10)
Scan100_Cockroach-2             49.4kB ± 0%    49.4kB ± 0%    ~             (p=0.157 n=8+9)
Scan1000_Cockroach-2             331kB ± 0%     331kB ± 0%  -0.01%        (p=0.005 n=10+10)
Scan10000_Cockroach-2           5.72MB ± 0%    5.72MB ± 0%    ~            (p=0.515 n=10+8)
Scan1000Limit1_Cockroach-2      14.1kB ± 0%    14.1kB ± 0%    ~            (p=0.062 n=10+9)
Scan1000Limit10_Cockroach-2     18.0kB ± 0%    17.9kB ± 0%  -0.19%         (p=0.013 n=9+10)
Scan1000Limit100_Cockroach-2    50.1kB ± 0%    50.1kB ± 0%    ~            (p=0.127 n=10+8)
PgbenchQuery_Cockroach-2         217kB ± 0%     216kB ± 0%  -0.37%         (p=0.000 n=8+10)

name                          old allocs/op  new allocs/op  delta
Bank_Cockroach-2                 1.67k ± 0%     1.67k ± 0%  -0.06%          (p=0.002 n=7+8)
Select1_Cockroach-2               97.0 ± 0%      97.0 ± 0%    ~     (all samples are equal)
Select2_Cockroach-2              2.70k ± 0%     2.70k ± 0%  -0.02%         (p=0.022 n=9+10)
Insert1_Cockroach-2                357 ± 0%       355 ± 0%  -0.78%        (p=0.000 n=10+10)
Insert10_Cockroach-2               917 ± 0%       904 ± 0%  -1.36%        (p=0.000 n=10+10)
Insert100_Cockroach-2            6.28k ± 0%     6.18k ± 0%  -1.60%          (p=0.000 n=9+9)
Insert1000_Cockroach-2           60.1k ± 0%     59.0k ± 0%  -1.73%        (p=0.000 n=10+10)
Update1_Cockroach-2                720 ± 0%       716 ± 0%  -0.57%         (p=0.000 n=8+10)
Update10_Cockroach-2             1.43k ± 0%     1.41k ± 0%  -1.55%         (p=0.000 n=9+10)
Update100_Cockroach-2            8.15k ± 0%     8.00k ± 2%  -1.84%        (p=0.000 n=10+10)
Update1000_Cockroach-2           72.4k ± 0%     70.4k ± 0%  -2.84%          (p=0.001 n=6+8)
Delete1_Cockroach-2                692 ± 0%       689 ± 0%  -0.49%         (p=0.000 n=10+7)
Delete10_Cockroach-2             1.72k ± 0%     1.70k ± 0%  -1.27%          (p=0.000 n=7+9)
Delete100_Cockroach-2            11.6k ± 0%     11.4k ± 0%  -1.74%          (p=0.000 n=9+9)
Delete1000_Cockroach-2            110k ± 0%      108k ± 0%  -1.82%          (p=0.000 n=9+8)
Scan1_Cockroach-2                  275 ± 1%       274 ± 1%  -0.33%        (p=0.006 n=10+10)
Scan10_Cockroach-2                 355 ± 0%       354 ± 0%  -0.23%        (p=0.019 n=10+10)
Scan100_Cockroach-2              1.08k ± 0%     1.08k ± 0%  -0.06%          (p=0.041 n=8+9)
Scan1000_Cockroach-2             8.29k ± 0%     8.29k ± 0%  -0.01%        (p=0.001 n=10+10)
Scan10000_Cockroach-2            80.4k ± 0%     80.4k ± 0%  -0.00%          (p=0.000 n=7+8)
Scan1000Limit1_Cockroach-2         298 ± 0%       298 ± 0%    ~            (p=0.161 n=9+10)
Scan1000Limit10_Cockroach-2        378 ± 0%       376 ± 0%  -0.32%        (p=0.001 n=10+10)
Scan1000Limit100_Cockroach-2     1.10k ± 0%     1.10k ± 0%    ~            (p=0.071 n=10+6)
PgbenchQuery_Cockroach-2         3.92k ± 0%     3.90k ± 0%  -0.62%         (p=0.000 n=8+10)
```

nvanbenschoten commented:
Yeah I agree that the allocations on every `heap.Pop` from the heap interface makes this alternative too unappealing to switch to. That said, I just came up with a mildly hacky way to avoid all allocations from the heap without needing to reimplement any stdlib functionality. See the second commit for perf numbers on this change.

If you don't think the hack is too gross, there are actually a few other places in our code that it could be applied to dramatically reduce allocations (like `baseQueue`). In some of those places, we could also make a similar optimization for `heap.Push` and `heap.Remove`.

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[storage/command_queue.go, line 227 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4878#-KC7Jm-2m_7oMN4f3ewI:-KC7Jm-2m_7oMN4f3ewJ:-141406353) ([raw file](https://github.com/cockroachdb/cockroach/blob/9b7151b31a3f97b023530c35372b39f734aea82b/storage/command_queue.go#L227)):*
might want to defer this

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4878#-:-KC7Jz2bm5ACin0qsFsd:1417088721)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[storage/command_queue.go, line 262 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4878#-KC7K63wakV4KuSqbmu-:-KC7K63wakV4KuSqbmu0:-1216610292) ([raw file](https://github.com/cockroachdb/cockroach/blob/9b7151b31a3f97b023530c35372b39f734aea82b/storage/command_queue.go#L262)):*
might want to `panic("unimplemented")` here

---

*[storage/command_queue.go, line 268 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4878#-KC7KP4p9WwcP71zjLod:-KC7KP4q-WzG6BwjqUnY:-1935535135) ([raw file](https://github.com/cockroachdb/cockroach/blob/9b7151b31a3f97b023530c35372b39f734aea82b/storage/command_queue.go#L268)):*
nit: don't need the `0`

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4878)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions.

---

*[storage/command_queue.go, line 227 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4878#-KC7Jm-2m_7oMN4f3ewI:-KC7QP7zaekwJhK-N3Qs:721887459) ([raw file](https://github.com/cockroachdb/cockroach/blob/9b7151b31a3f97b023530c35372b39f734aea82b/storage/command_queue.go#L227)):*
We're in a loop here.

---

*[storage/command_queue.go, line 262 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4878#-KC7K63wakV4KuSqbmu-:-KC7QOd44Zm4H-THXuCg:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/9b7151b31a3f97b023530c35372b39f734aea82b/storage/command_queue.go#L262)):*
Done.

---

*[storage/command_queue.go, line 268 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4878#-KC7KP4p9WwcP71zjLod:-KC7QJnW3nn1X9ERrXX8:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/9b7151b31a3f97b023530c35372b39f734aea82b/storage/command_queue.go#L268)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4878)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[storage/command_queue.go, line 227 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4878#-KC7Jm-2m_7oMN4f3ewI:-KC7R8_0IGWax5pQyNwR:457472232) ([raw file](https://github.com/cockroachdb/cockroach/blob/9b7151b31a3f97b023530c35372b39f734aea82b/storage/command_queue.go#L227)):*
OK. Seems like you could defer it outside the loop.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4878)*
<!-- Sent from Reviewable.io -->

petermattis commented:
LGTM, I don't think the hack is gross or even much of a hack. 

---

Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[storage/command_queue.go, line 227 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4878#-KC7Jm-2m_7oMN4f3ewI:-KC7e8WtugpAfVKT00LX:-1811505798) ([raw file](https://github.com/cockroachdb/cockroach/blob/9b7151b31a3f97b023530c35372b39f734aea82b/storage/command_queue.go#L227)):*
I think the heap needs to be cleared on every iteration. Perhaps for safety you could defer outside the loop as well, but there are no early returns and we'll crash on a panic so I'm not sure what that is buying you.

---

*[storage/command_queue.go, line 267 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4878#-KC7faNz3DGJwf-8jrR2:-KC7faNz3DGJwf-8jrR3:1403889589) ([raw file](https://github.com/cockroachdb/cockroach/blob/47955440cb644c952b0f9b8d2bf0c1fb77e27957/storage/command_queue.go#L267)):*
Add a comment (similar to what is in the commit message) that you're returning a pointer to avoid an allocation when storing the `cache.Overlap` in an `interface{}`. 

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4878#-:-KC7g7kIBSjQrZ3sGYPf:945613045)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
LGTM. I also don't think it's really a hack, just something subtle that needs to be commented.

---

Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[storage/command_queue.go, line 55 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4878#-KC8-gobN_IbegEJFeCU:-KC8-goctHSaChinYEti:-325780578) ([raw file](https://github.com/cockroachdb/cockroach/blob/47955440cb644c952b0f9b8d2bf0c1fb77e27957/storage/command_queue.go#L55)):*
Tiny grammar nit: `wRg, rwRg` is plural, `oHeap` is singular, so if only one of them has an `s` on `avoids`, it should be the latter.  Using the same comment for both lines (with or without the `s`) would also be fine.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4878#-:-KC8-MoGRJGnr0x7-fcU:431116310)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Btw, recent changes that just went into the photos app will totally change the character of your benchmarks going forward, at least early on in the process. There will be way more contention with high concurrency because many of the ops end up choosing a random user which already exists now that we have an efficient limit in sql. Just a heads up, and perhaps we should change the code to work around that.

nvanbenschoten commented:
Oh don't worry, the hack was much worse before @tamird reminded me that we could address directly into the slice (as long as we deference ASAP).

And thanks for the heads up Spencer, I'll give the new photos app another look under a profiler and see if anything pops out.

---

Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions.

---

*[storage/command_queue.go, line 55 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4878#-KC8-gobN_IbegEJFeCU:-KCGfaHuX8M0Ja99689I:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/47955440cb644c952b0f9b8d2bf0c1fb77e27957/storage/command_queue.go#L55)):*
Done.

---

*[storage/command_queue.go, line 267 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4878#-KC7faNz3DGJwf-8jrR2:-KCGgbtMRMNSjmQmE7vG:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/47955440cb644c952b0f9b8d2bf0c1fb77e27957/storage/command_queue.go#L267)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4878#-:-KCGew8C8NRmSNM9I95L:-425927564)*
<!-- Sent from Reviewable.io -->
